### DISCUSSION TITLE
Get raw DLPack capsule from nb::ndarray

### DISCUSSION THREAD

**Discussion by awni** [wjakob/nanobind#583] [Q&A]
From the docs I thought the following would work (e.g. no framework annotation):

```c++
      const float data[] = { 1, 2, 3, 4, 5, 6, 7, 8 };
      m.def("ret_dlpack", [](const array& a) {         
        size_t shape[2] = { 2, 4 };
        return nb::ndarray<const float, nb::shape<2, -1>>(
            /* data = */ data,                       
            /* ndim = */ 2,                          
            /* shape pointer = */ shape,             
            /* owner = */ nb::handle());             
      })    
```

But it fails with the error:

```
RuntimeError: nanobind::detail::ndarray_wrap(): copy failed: AttributeError: 'PyCapsule' object has no attribute 'copy'
```

Is that expected? If so, is it possible to access and return the DLPack capsule in a bound function (e.g. `__dlpack__`).

**Comment by wjakob** [comment]
The error message is weird, I agree with that. I think the key problem is that you didn't specify an owner, which means that nanobind must make a copy. But it cannot make a copy of a raw DLPack capsule because the data might be wherever (e.g. on a CUDA or ROCm device, and nanobind doesn't link against these libraries. So we cannot memcpy). What happens when you set the `owner` argument to `nb::none()`?

**Reply by awni** [reply]
Indeed it runs without the error if I set the owner to `nb::none()`. Thanks!

In the actual use case the data pointer is coming from an MLX array. Presumably the `nb::ndarray` will not copy the data. If that's correct, then I should use `nb::rv_policy::reference_internal`  as the return value policy? That way the capsule keeps the array alive. When the capsule is dereferenced, the array releases the data. Does that sound correct to you?
